source/game/system/map/MapdataCheckPoint.cc

```diff
@@ -18,17 +21,115 @@ MapdataCheckPoint::MapdataCheckPoint(const SData *data)
 void MapdataCheckPoint::read(EGG::Stream &stream) {
     m_left.read(stream);
     m_right.read(stream);
-    m_jugemIndex = stream.read_u8();
+    m_jugemIndex = stream.read_s8();
     m_checkArea = stream.read_s8();
-    m_prevPt = stream.read_u8();
-    m_nextPt = stream.read_u8();
+    m_prevPt = stream.read_s8();
+    m_nextPt = stream.read_s8();
+}
+
+/// @brief Calculates @ref m_nextPoints and @ref m_prevPoints from @ref m_nextPt and @ref m_prevPt.
+/// @details Also calculates the quadrilaterals for the next checkpoints, filling the fields of @ref
+/// LinkedCheckpoint for each.
+/// @addr{0x80515624}
+void MapdataCheckPoint::initCheckpointLinks(MapdataCheckPointAccessor &accessor, int id) {
+    m_id = id;
+    auto checkPathAccessor = CourseMap::Instance()->getCheckPathAccessor();
+
+    // Calculate the quadrilateral's `m_prevPoints`. If the check point is the first in its group,
+    // it has multiple previous checkpoints defined by its preceding checkpaths
+    if (m_prevPt == -1) {
+        MapdataCheckPath *checkpath = checkPathAccessor->findCheckpathForCheckpoint(id);
+        m_prevCount = 0;
+
+        for (auto [i, prevID] : std::views::enumerate(checkpath->prev())) {
+            if (prevID == 0xff) {
+                continue;
+            }
```

Comment:
Probably want to change `prevID` to `s8` to that we can more clearly write `prevID == -1`